### ISSUE TITLE
Wrong suggestion of `const fn` if member of member implements Drop

### ISSUE THREAD

**Issue by ArekPiekarz** [rust-lang/rust-clippy#4979]
cargo clippy -V: clippy 0.0.212 (c807fbc 2019-12-29)

### Summary
A warning "this could be a const_fn" is triggered for a function of a struct  that doesn't implement Drop, but on some level of depth contains a member that does implement it. Adding const to such function results in compilation error.

### Steps to reproduce 

Run the following code with `cargo clippy -- -W clippy::nursery`:
```rust
#![allow(clippy::blacklisted_name)]
#![allow(dead_code)]
#![allow(unused_variables)]

struct Foo
{
    field: String
}

impl Foo
{
    fn take(self) -> String
    {
        self.field
    }
}

fn main()
{
}
```
It results in a warning:
```
warning: this could be a const_fn
  --> src/main.rs:12:5
   |
12 | /     fn take(self) -> String
13 | |     {
14 | |         self.field
15 | |     }
   | |_____^
   |
   = note: `-W clippy::missing-const-for-fn` implied by `-W clippy::nursery`
   = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
```
After adding const to the function `Foo::take` we get an error:
```
error[E0493]: destructors cannot be evaluated at compile-time
  --> src/main.rs:12:19
   |
12 |     const fn take(self) -> String
   |                   ^^^^ constant functions cannot evaluate destructors
```
Note that Foo doesn't implement Drop, its field String also doesn't, but String's field Vec<u8> does implement it.
Also note there was already a similar fixed issue, but for structs that implement Drop: https://github.com/rust-lang/rust-clippy/issues/4449

**Comment by expenses** [comment]
This is now triggering more because more things were made const in rust 1.61.0.

naga has contains the following code:

```rust

impl<W: Write> Writer<W> {
    /// Finishes writing and returns the output.
    pub fn finish(self) -> W {
        self.out
    }
...
}
```

This triggers the lint:

```
warning: this could be a `const fn`
   --> src/back/msl/writer.rs:650:5
    |
650 | /     pub fn finish(self) -> W {
651 | |         self.out
652 | |     }
    | |_____^
    |
note: the lint level is defined here
   --> src/lib.rs:199:5
    |
199 |     clippy::missing_const_for_fn
    |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    = help: for further information visit https://rust-lang.github.io/rust-clippy/master/index.html#missing_const_for_fn
```

but when you actually apply that you get an error:
```
error[E0493]: destructors cannot be evaluated at compile-time
   --> src/back/msl/writer.rs:650:25
    |
650 |     pub const fn finish(self) -> W {
    |                         ^^^^ constant functions cannot evaluate destructors
651 |         self.out
652 |     }
    |     - value is dropped here

For more information about this error, try `rustc --explain E0493`.
```

**Comment by samueltardieu** [comment]
This seems fixed. Can we close the issue?
